### PR TITLE
Rename Buffer Functions

### PR DESCRIPTION
**Pull request by kkoopa**
These fit the pattern better. Apparently there were never any tests of buffer functionality.

### CODE REVIEW

**Comment by bnoordhuis** [comment]
The CI doesn't seem particularly thrilled.  Is this a breaking change?

Apropos the `s/NanBufferUse/NanBufferNew/` change: `Buffer::Use()` and `Buffer::New()` do different things.  The former takes ownership of the pointer, the latter makes a copy (unless you use the `Buffer::New()` overload that takes a callback.)

**Comment by kkoopa** [comment]
CI is broken because of a yet unpublished tap update.

There's a comment in the buffer source stating that it really should be called Buffer::New, figured it might as well be.

**Comment by kkoopa** [comment]
@trevnorris Re: https://github.com/nodejs/io.js/blob/master/src/node_buffer.h#L63 Is this still the case?

**Comment by kkoopa** [comment]
Actually, if we go with #356, we could introduce namespace Nan::Buffer to mirror node::Buffer, minus the isolates, or why not use a class?

**Comment by trevnorris** [comment]
@kkoopa Neither of those are actually going to happen. Would cause too much breakage.

**Comment by kkoopa** [comment]
Yes, but I'm willing to do that in Nan. Should it be done?

**Comment by trevnorris** [comment]
It's just a name switch to make the API make more sense. So completely aesthetic.

**Comment by kkoopa** [comment]
I'll take that as a 'yes'.

**Comment by kkoopa** [comment]
Rebased

**Comment by bnoordhuis** [comment]
LGTM.  I'm kind of sad there are no tests for the NanBuffer functions.

**Comment by kkoopa** [comment]
I'll add some tests tomorrow and merge when CI is happy. It actually seems to error out on 0.8 right now: https://travis-ci.org/nodejs/nan/jobs/64168203

**Comment by trevnorris** [comment]
@kkoopa a minor API change is that `node::Buffer::New()` used to abort if the length passed was `> 0x3fffffff`. Once Buffers are backed by Uint8Array you'll be able to pass an arbitrary size. Thing is, if `malloc()` returns `nullptr` then `Buffer::New()` will return an empty object. Indicating that the memory couldn't be allocated.

don't know if this will affect nan, but just in case wanted to give you a heads up.

**Comment by kkoopa** [comment]
As we even support node 0.8, that limit will stay in NAN on all platforms to make it consistent. I guess that assert should be added to the NAN wrapper then.

**Comment by trevnorris** [comment]
Sad that the new functionality will be missing, but understand where you're coming from.

**Comment by kkoopa** [comment]
If someone wants at it, they can always add alternate code paths with ifdefs themselves.

**Comment by kkoopa** [comment]
Added tests,

**Comment by trevnorris** [comment]
@kkoopa Upon further implementation I realized that having the length be a `uint32_t` isn't sufficient. It will need to change to `size_t`. I'm unsure whether to change the existing API or create a new one that accepts `size_t`. Either way, wanted to let you know.

**Comment by kkoopa** [comment]
You can safely change the existing API to use `size_t`, since it is bigger than `uint32_t` and has the same sign. Upcasting is done automatically. NAN will keep `uint32_t` for backwards compatibility. I'll add `isEmpty` checks.

**Comment by kkoopa** [comment]
Where's [this test failure](https://travis-ci.org/nodejs/nan/jobs/64180954) in Buffer coming from?

**Comment by kkoopa** [comment]
I guess it's because `Buffer::Use` tries to free the supplied buffer, which does not work, as it is static.

**Comment by kkoopa** [comment]
@trevnorris, if there is a need to do an `IsEmpty` check, shouldn't `Buffer` be rewritten to return a `MaybeLocal<Object>`?

File: nan.h [hunk 1]
```diff
@@ -605,31 +605,45 @@ class NanEscapableScope {
...
 
-  NAN_INLINE v8::Local<v8::Object> NanBufferUse(
+  NAN_INLINE v8::Local<v8::Object> NanNewBuffer(
       char* data
     , uint32_t size
   ) {
+    // arbitrary buffer lengths requires
+    // NODE_MODULE_VERSION >= IOJS_3_0_MODULE_VERSION
+    assert(size <= 0x3fffffff && "too large buffer");
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
```
Comment (kkoopa) [inline nan.h]: I'm assuming this one cannot return an empty handle, as it should not be calling malloc.

File: nan.h [hunk 2]
```diff
@@ -1008,10 +1031,13 @@ class NanEscapableScope {
     delete[] data;
```
Comment (kkoopa) [inline nan.h]: Is this correct? I though smalloc used `malloc`, which means that this should be `free(data)`.

**Comment by trevnorris** [comment]
Yes. `Buffer::Use()` assumes it has control of the memory that's passed to it.

I'm not sure about returning a `MaybeLocal<>`. Going to get some feedback about that.

File: nan.h (hunk 1 above)
Comment (trevnorris) [inline nan.h]: You're correct.

**Comment by kkoopa** [comment]
`Buffer` was updated to take `Isolate *` arguments when that change hit V8, this would be the natural thing to do for following in that line, especially since nobody will expect that a `Local` can be empty any more since V8 4.3, as all those APIs return a `MaybeLocal` in those cases. Not returning `MaybeLocal` from `Buffer` would thus be error-prone and unnatural.

**Comment by trevnorris** [comment]
By "not sure" I meant that I wasn't sure what the other maintainers thought about the idea of changing the return value. After brief discussion it does seem returning `MaybeLocal` is the smart thing to do.

**Comment by kkoopa** [comment]
Merged.

**Pull request closed without merging**